nixos/modules/services/networking/dns-over-https.nix

```diff
@@ -0,0 +1,193 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.dns-over-https;
+  clientBootstrap = concatMapStringsSep "," (x: "\"${x}\"") cfg.client.bootstrap;
+  clientListen = concatMapStringsSep "," (x: "\"${x}\"") cfg.client.listen;
+  clientUpstream = concatMapStrings (s: "[[upstream.upstream_ietf]]\n  url = \"${s.url}\"\n  weight = ${toString s.weight}\n") cfg.client.upstream;
+  serverListen = concatMapStringsSep "," (x: "\"${x}\"") cfg.server.listen;
+  serverUpstream = concatMapStringsSep "," (x: "\"${x}\"") cfg.server.upstream;
+
+  # See upstream: https://github.com/m13253/dns-over-https/blob/master/doh-client/doh-client.conf
+  clientConf = pkgs.writeText "doh-client.conf" ''
+    listen = [${clientListen}]
+
+    [upstream]
+    upstream_selector = "${cfg.client.upstreamSelector}"
+
+    ${clientUpstream}
+
+    [others]
+    bootstrap = [${clientBootstrap}]
+    passthrough = [
+        "captive.apple.com",
+        "connectivitycheck.gstatic.com",
+        "detectportal.firefox.com",
+        "msftconnecttest.com",
+        "nmcheck.gnome.org",
+        "pool.ntp.org",
+        "time.apple.com",
+        "time.asia.apple.com",
+        "time.euro.apple.com",
+        "time.nist.gov",
+        "time.windows.com",
+    ]
+    timeout = 30
+    no_cookies = true
+    no_ecs = false
+    no_ipv6 = false
+    no_user_agent = false
+    verbose = ${boolToString cfg.verbose}
+  '';
+  # See upstream: https://github.com/m13253/dns-over-https/blob/master/doh-server/doh-server.conf
+  serverConf = pkgs.writeText "doh-server.conf" ''
+    listen = [${serverListen}]
+    local_addr = ""
+    cert = ""
+    key = ""
+    path = "/dns-query"
+    upstream = [${serverUpstream}]
+    timeout = 10
+    tries = 3
+    verbose = ${boolToString cfg.verbose}
+    log_guessed_client_ip = false
+  '';
+
+in
+{
+  options = {
+    services.dns-over-https = {
+      verbose = mkOption {
+        type = types.bool;
+        default = false;
+        description = "Enable logging";
+      };
+      client = mkOption {
+        description = "DNS-over-HTTPS Client";
+        type = types.submodule {
+          options = {
+            enable = mkEnableOption "DNS-over-HTTPS Client";
+            bootstrap = mkOption {
+              description = "Bootstrap DNS server to resolve the address of the upstream resolver";
+              type = types.listOf types.str;
+              default = [ "8.8.8.8:53" "8.8.4.4:53" ];
+              example = [
+                # Google's resolver, bad ECS, good DNSSEC
+                "8.8.8.8:53"
+                "8.8.4.4:53"
```

Comment:
Bad ECS here?  That sounds wrong... or it needs more explanation.  I'm fairly confident that this pair of Google addresses has good [client subnet](https://www.rfc-editor.org/rfc/rfc7871.html) support over port 53.

Reply:
Yeah, the example config upstream seems to contradict itself, likely a mistake:

https://github.com/m13253/dns-over-https/blob/b74220718f8ed0e09f2f2f09205039af556845e1/doh-client/doh-client.conf#L21 vs https://github.com/m13253/dns-over-https/blob/b74220718f8ed0e09f2f2f09205039af556845e1/doh-client/doh-client.conf#L67